core/src/unprocessed_packet_batches.rs

```diff
@@ -325,27 +330,64 @@ pub fn transactions_to_deserialized_packets(
 // This function deserializes packets into transactions, computes the blake3 hash of transaction
 // messages, and verifies secp256k1 instructions. A list of sanitized transactions are returned
 // with their packet indexes.
-#[allow(clippy::needless_collect)]
 pub fn transaction_from_deserialized_packet(
     deserialized_packet: &ImmutableDeserializedPacket,
     feature_set: &Arc<feature_set::FeatureSet>,
     votes_only: bool,
     address_loader: impl AddressLoader,
+    tx_account_lock_limit: usize,
+    transaction_error_counters: &mut TransactionErrorMetrics,
 ) -> Option<SanitizedTransaction> {
     if votes_only && !deserialized_packet.is_simple_vote() {
         return None;
     }
-    let tx = SanitizedTransaction::try_new(
-        deserialized_packet.transaction().clone(),
-        *deserialized_packet.message_hash(),
-        deserialized_packet.is_simple_vote(),
+
+    // Check that the account locks are valid
+    let message = SanitizedMessage::try_new(
+        deserialized_packet.transaction().get_message().clone(),
         address_loader,
     )
     .ok()?;
+    match SanitizedTransaction::validate_account_locks(&message, tx_account_lock_limit) {
+        Err(TransactionError::AccountLoadedTwice) => {
+            saturating_add_assign!(transaction_error_counters.total, 1);
+            saturating_add_assign!(transaction_error_counters.account_loaded_twice, 1);
+        }
+        Err(TransactionError::TooManyAccountLocks) => {
+            saturating_add_assign!(transaction_error_counters.total, 1);
+            saturating_add_assign!(transaction_error_counters.too_many_account_locks, 1);
+        }
+        _ => {}
+    }
+
+    let tx = SanitizedTransaction::new(
+        deserialized_packet.transaction().clone(),
+        message,
+        *deserialized_packet.message_hash(),
+        deserialized_packet.is_simple_vote(),
+    );
```

Comment:
Not sure if there is substantial benefit of creating sanitized_message separately, but Ln351 uses SanitizedTransaction's function to validate a sanitizedMessage seems to break encapsulation; Also somewhat uneasy about Ln363 to allow creating sanitized_transaction with a transaction and a potentially unrelated sanitized_message. 

Why not something like this: 
```
let tx = SanitizedTransaction::try_new(...).ok()?; // existing code
tx.verify_precompiles(feature_set).ok()?; // existing code
tx.verify_account_locks(tx_account_lock_limit).ok()?; // new public function from sanitized_transaction
...
}
```

Reply:
I think my thought was we avoid cloning the transaction until we know we have valid locks. I don't think that's a big concern; we should keep the code cleaner!